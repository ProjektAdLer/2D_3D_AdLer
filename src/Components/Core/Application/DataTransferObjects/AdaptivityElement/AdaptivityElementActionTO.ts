export default class AdaptivityElementActionTO {
  actionType: string;
  actionData: string | number;
}
