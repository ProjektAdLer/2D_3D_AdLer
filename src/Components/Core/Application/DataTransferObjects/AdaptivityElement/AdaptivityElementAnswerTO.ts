export default class AdaptivityElementAnswersTO {
  answerId: number;
  answerText?: string;
  answerImage?: string;
}
