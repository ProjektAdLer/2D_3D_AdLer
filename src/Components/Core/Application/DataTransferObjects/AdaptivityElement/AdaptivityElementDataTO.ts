import AdaptivityElementTaskTO from "./AdaptivityElementTaskTO";

export class AdaptivityElementDataTO {
  introText: string;
  shuffleTask: boolean;
  tasks: AdaptivityElementTaskTO[];
}
