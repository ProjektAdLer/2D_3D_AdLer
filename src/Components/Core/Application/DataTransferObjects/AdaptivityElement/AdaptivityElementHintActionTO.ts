import { AdaptivityElementActionTypes } from "src/Components/Core/Domain/Types/Adaptivity/AdaptivityElementActionTypes";

// similar structure as AdaptivityElementViewModel
export default class AdaptivityElementHintActionTO {
  hintActionType: AdaptivityElementActionTypes;
  idData?: number;
  textData?: string;
}
