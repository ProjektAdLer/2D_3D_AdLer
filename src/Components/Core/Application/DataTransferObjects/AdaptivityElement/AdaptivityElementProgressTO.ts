import { AdaptivityDataTO } from "../AdaptivityDataTO";
import AdaptivityElementTaskProgressTO from "./AdaptivityElementTaskProgressTO";

export default class AdaptivityElementProgressTO extends AdaptivityDataTO {
  isCompleted: boolean;
  override tasks: AdaptivityElementTaskProgressTO[];
}
