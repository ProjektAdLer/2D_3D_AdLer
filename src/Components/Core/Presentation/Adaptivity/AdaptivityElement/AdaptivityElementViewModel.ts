import { AdaptivityElementQuestionDifficultyTypes } from "src/Components/Core/Domain/Types/Adaptivity/AdaptivityElementQuestionDifficultyTypes";
import Observable from "src/Lib/Observable";
import { StyledButtonColor } from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";

export interface AdaptivityAnswer {
  answerIndex: number;
  answerText: string;
  isSelected: boolean;
}

export interface AdaptivityQuestion {
  questionID: number;
  questionText: string;
  questionAnswers: AdaptivityAnswer[];
  isRequired: boolean;
  isCompleted: boolean;
  difficulty: AdaptivityElementQuestionDifficultyTypes;
  isMultipleChoice: boolean;
}

export interface AdaptivityTask {
  taskID: number;
  taskTitle: string;
  questions: AdaptivityQuestion[];
  isCompleted: boolean;
  requiredDifficulty: number;
}

export interface AdaptivityElementContent {
  elementName: string;
  introText: string;
  tasks: AdaptivityTask[];
}

export default class AdaptivityElementViewModel {
  isOpen: Observable<boolean> = new Observable<boolean>(false);
  showFeedback: Observable<boolean> = new Observable<boolean>(false);
  contentData: Observable<AdaptivityElementContent> =
    new Observable<AdaptivityElementContent>();
  currentTaskID: Observable<number | null> = new Observable<number | null>(
    null
  );
  currentQuestionID: Observable<number | null> = new Observable<number | null>(
    null
  );
  footerText: Observable<string> = new Observable<string>("");
}
