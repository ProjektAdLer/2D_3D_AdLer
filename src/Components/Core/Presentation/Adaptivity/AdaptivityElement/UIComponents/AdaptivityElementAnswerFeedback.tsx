import { useEffect, useRef } from "react";
import { useTranslation } from "react-i18next";
import StyledButton from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledButton";
import {
  AdaptivityQuestion,
  AdaptivityTask,
} from "../AdaptivityElementViewModel";

export default function AdaptivityElementAnswerFeedback({
  isCorrect,
  currentQuestion,
  currentTask,
  setHeaderText,
  closeFeedback,
}: {
  isCorrect: boolean;
  currentQuestion: AdaptivityQuestion;
  currentTask: AdaptivityTask;
  setHeaderText: (text: string) => void;
  closeFeedback: () => void;
}) {
  const { t: translate } = useTranslation("learningElement");
  useEffect(() => {
    setHeaderText(
      isCorrect ? translate("rightAnswer") : translate("wrongAnswer")
    );
  }, [isCorrect, setHeaderText, translate]);
  const difficultyHint = useRef(null as string | null);
  useEffect(() => {
    currentTask.questions.some((question) => {
      return (
        question.difficulty.valueOf() > currentQuestion.difficulty.valueOf()
      );
    }) &&
      isCorrect &&
      (difficultyHint.current = translate(
        "questionSolvedWithHigherDifficultyQuestionAvailable"
      ).toString());

    currentTask.questions.some((question) => {
      return (
        question.difficulty.valueOf() < currentQuestion.difficulty.valueOf()
      );
    }) &&
      !isCorrect &&
      (difficultyHint.current = translate(
        "questionSolvedWithLowerDifficultyQuestionAvailable"
      ));
  }, [currentQuestion, currentTask, isCorrect, translate]);
  const solvedHint = useRef(null as string | null);
  useEffect(() => {
    if (
      // hint is given if the solved question is required or if a question of lower difficulty is required
      (currentTask.questions.some((question) => {
        return (
          question.isRequired &&
          question.difficulty.valueOf() < currentQuestion.difficulty.valueOf()
        );
      }) ||
        currentQuestion.isRequired) &&
      isCorrect
    ) {
      solvedHint.current = translate("requiredQuestionSolvedCorrectly");
    }
  }, [
    currentQuestion.difficulty,
    currentQuestion.isRequired,
    currentTask,
    isCorrect,
    translate,
  ]);

  return (
    <div className="flex flex-col gap-4 pl-4 pr-2 my-4 h-fit">
      <div className="flex flex-col items-start justify-start bg-buttonbgblue rounded-xl">
        {solvedHint.current && <div className="p-2">{solvedHint.current}</div>}
        {difficultyHint.current && (
          <div className="p-2">{difficultyHint.current}</div>
        )}
      </div>
      <div className="flex justify-end w-full">
        <StyledButton shape="freefloatcenter" onClick={closeFeedback}>
          <p className="text-sm">{translate("nextButton")}</p>
        </StyledButton>
      </div>
    </div>
  );
}
