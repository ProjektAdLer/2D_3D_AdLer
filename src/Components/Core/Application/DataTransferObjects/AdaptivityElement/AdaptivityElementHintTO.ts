import { AdaptivityHintAction } from "src/Components/Core/Presentation/Adaptivity/AdaptivityElement/AdaptivityElementViewModel";

// similar structure as in AdaptivityElementViewModel
export default class AdaptivityElementHintTO {
  hintID: number;
  showOnIsWrong: boolean;
  hintAction: AdaptivityHintAction;
}
