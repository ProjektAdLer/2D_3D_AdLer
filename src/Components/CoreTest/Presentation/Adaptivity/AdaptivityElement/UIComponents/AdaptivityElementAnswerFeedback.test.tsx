import { render, waitFor } from "@testing-library/react";
import React from "react";
import AdaptivityElementAnswerFeedback from "../../../../../Core/Presentation/Adaptivity/AdaptivityElement/UIComponents/AdaptivityElementAnswerFeedback";
import { AdaptivityQuestion } from "../../../../../Core/Presentation/Adaptivity/AdaptivityElement/AdaptivityElementViewModel";

const easyQuestion: AdaptivityQuestion = {
  questionID: 0,
  questionText: "testQuestionText",
  isMultipleChoice: false,
  difficulty: 0,
  isCompleted: false,
  isRequired: false,
  questionAnswers: [
    {
      answerIndex: 0,
      answerText: "testAnswerText",
      isSelected: false,
    },
  ],
  hints: [],
};
const middleQuestion: AdaptivityQuestion = {
  questionID: 1,
  questionText: "testQuestionText",
  isMultipleChoice: false,
  difficulty: 100,
  isCompleted: false,
  isRequired: false,
  questionAnswers: [
    {
      answerIndex: 0,
      answerText: "testAnswerText",
      isSelected: false,
    },
  ],
  hints: [],
};
const hardQuestion: AdaptivityQuestion = {
  questionID: 2,
  questionText: "testQuestionText",
  isMultipleChoice: false,
  difficulty: 200,
  isCompleted: false,
  isRequired: false,
  questionAnswers: [
    {
      answerIndex: 0,
      answerText: "testAnswerText",
      isSelected: false,
    },
  ],
  hints: [],
};

describe("AdaptivityElementAnswerFeedback", () => {
  test("should render", () => {
    const { container } = render(
      <AdaptivityElementAnswerFeedback
        isCorrect={false}
        currentQuestion={easyQuestion}
        currentTask={{
          taskID: 0,
          taskTitle: "testTaskTitle",
          isCompleted: false,
          isRequired: false,
          hasBeenCompleted: false,
          requiredDifficulty: 0,
          questions: [easyQuestion],
        }}
        setHeaderText={() => {}}
        closeFeedback={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  // ANF-ID: [EWE0016]
  test("answering a question correctly and there exists a more difficult question then a more difficult question will be recommended", () => {
    const container = render(
      <AdaptivityElementAnswerFeedback
        isCorrect={true}
        currentQuestion={easyQuestion}
        currentTask={{
          taskID: 0,
          taskTitle: "testTaskTitle",
          isCompleted: false,
          isRequired: false,
          hasBeenCompleted: false,
          requiredDifficulty: 0,
          questions: [easyQuestion, middleQuestion, hardQuestion],
        }}
        setHeaderText={() => {}}
        closeFeedback={() => {}}
      />
    );

    waitFor(() => {
      expect(
        container.getByText(
          "questionSolvedWithHigherDifficultyQuestionAvailable"
        )
      ).toBeInTheDocument();
    });
  });

  test("answering a question correctly and there exists no question that is more difficult then no more difficult question will be recommended", () => {
    const container = render(
      <AdaptivityElementAnswerFeedback
        isCorrect={true}
        currentQuestion={middleQuestion}
        currentTask={{
          taskID: 0,
          taskTitle: "testTaskTitle",
          isCompleted: false,
          isRequired: false,
          hasBeenCompleted: false,
          requiredDifficulty: 0,
          questions: [easyQuestion, middleQuestion],
        }}
        setHeaderText={() => {}}
        closeFeedback={() => {}}
      />
    );

    waitFor(() => {
      expect(
        container.getByText(
          "questionSolvedWithHigherDifficultyQuestionAvailable"
        )
      ).not.toBeInTheDocument();
    });
  });

  // ANF-ID: [EWE0016]
  test("answering a question incorrectly and there exists a less difficult question then a less difficult question will be recommended", () => {
    const container = render(
      <AdaptivityElementAnswerFeedback
        isCorrect={false}
        currentQuestion={middleQuestion}
        currentTask={{
          taskID: 0,
          taskTitle: "testTaskTitle",
          isCompleted: false,
          isRequired: false,
          hasBeenCompleted: false,
          requiredDifficulty: 0,
          questions: [easyQuestion, middleQuestion, hardQuestion],
        }}
        setHeaderText={() => {}}
        closeFeedback={() => {}}
      />
    );

    waitFor(() => {
      expect(
        container.getByText(
          "questionSolvedWithLowerDifficultyQuestionAvailable"
        )
      ).toBeInTheDocument();
    });
  });

  test("answering a question incorrectly and there exists no less difficult question then no less question will be recommended", () => {
    const container = render(
      <AdaptivityElementAnswerFeedback
        isCorrect={false}
        currentQuestion={middleQuestion}
        currentTask={{
          taskID: 0,
          taskTitle: "testTaskTitle",
          isCompleted: false,
          isRequired: false,
          hasBeenCompleted: false,
          requiredDifficulty: 0,
          questions: [middleQuestion, hardQuestion],
        }}
        setHeaderText={() => {}}
        closeFeedback={() => {}}
      />
    );

    waitFor(() => {
      expect(
        container.getByText(
          "questionSolvedWithLowerDifficultyQuestionAvailable"
        )
      ).not.toBeInTheDocument();
    });
  });
});
