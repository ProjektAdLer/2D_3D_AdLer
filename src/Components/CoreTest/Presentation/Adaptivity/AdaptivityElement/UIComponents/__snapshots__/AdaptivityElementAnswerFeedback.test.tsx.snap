// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdaptivityElementAnswerFeedback should render 1`] = `
<div>
  <div
    class="flex flex-col gap-4 pl-4 pr-2 my-4 h-fit"
  >
    <div
      class="flex flex-col items-start justify-start bg-buttonbgblue rounded-xl"
    />
    <div
      class="flex justify-end w-full"
    >
      <div
        class=""
      >
        <button
          class="px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer hover:bg-adleryellow lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"
        >
          <p
            class="text-sm"
          >
            nextButton
          </p>
        </button>
      </div>
    </div>
  </div>
</div>
`;
